src/v/io/cache.h

```diff
@@ -0,0 +1,662 @@
+    if (small_queue_size_ > max_small_queue_size_) {
+        return evict_small();
+    }
+    return evict_main();
+}
+
+template<
+  typename T,
+  cache_hook T::*Hook,
+  cache_evictor<T> Evictor,
+  cache_cost<T> Cost>
+struct cache<T, Hook, Evictor, Cost>::stat
+cache<T, Hook, Evictor, Cost>::stat() const noexcept {
+    return {
+      .small_queue_size = small_queue_size_,
+      .main_queue_size = main_queue_size_,
+    };
+}
+
+template<
+  typename T,
+  cache_hook T::*Hook,
+  cache_evictor<T> Evictor,
+  cache_cost<T> Cost>
+bool cache<T, Hook, Evictor, Cost>::on_ghost_queue(
+  const T& entry) const noexcept {
+    /*
+     * if the insertion time is unspecified then the entry is not on the ghost
+     * queue. for example, it was evicted in insert, or has default-init hook.
+     * otherwise, the entry is on the ghost queue if it hasn't yet "timed out":
+     *
+     *    expires = insertion_time + main queue capacity
+     *        now = ghost_inserted
+     *   on_ghost = expires > now
+     */
+    const auto& hook = entry.*Hook;
+    if (hook.insertion_time_.has_value()) {
+        auto expires = *hook.insertion_time_ + max_main_queue_size_;
+        return expires > ghost_queue_age_;
+    }
+    return false;
+}
+
+template<
+  typename T,
+  cache_hook T::*Hook,
+  cache_evictor<T> Evictor,
+  cache_cost<T> Cost>
+bool cache<T, Hook, Evictor, Cost>::evict_small() noexcept {
+    for (auto it = small_fifo_.begin(); it != small_fifo_.end();) {
+        auto& entry = *it;
+        auto& hook = entry.*Hook;
+        assert(hook.location_ == cache_hook::location::small);
+        if (hook.freq_ > 1) {
+            hook.freq_ = 0;
+
+            // evict from small queue
+            it = small_fifo_.erase(it);
+            const auto cost = cost_(entry);
+            small_queue_size_ -= cost;
+
+            // promote to main queue
+            main_fifo_.push_back(entry);
+            hook.location_ = cache_hook::location::main;
+            main_queue_size_ += cost;
+
+            if (main_queue_size_ > max_main_queue_size_) {
+                if (evict_main()) {
+                    return true;
+                }
+            }
+
+        } else if (evict_(entry)) {
+            // evict from small queue
+            it = small_fifo_.erase(it);
+            const auto cost = cost_(entry);
+            small_queue_size_ -= cost;
+
+            // insert into ghost queue
+            hook.insertion_time_ = ghost_queue_age_;
+            ghost_queue_age_ += cost;
+            return true;
+
+        } else {
+            ++it;
+        }
+    }
+    return false;
+}
+
+template<
+  typename T,
+  cache_hook T::*Hook,
+  cache_evictor<T> Evictor,
+  cache_cost<T> Cost>
+bool cache<T, Hook, Evictor, Cost>::evict_main() noexcept {
+    for (auto it = main_fifo_.begin(); it != main_fifo_.end();) {
+        auto& entry = *it;
+        auto& hook = entry.*Hook;
+        assert(hook.location_ == cache_hook::location::main);
```

Comment:
why normal assert and not vassert?

Reply:
i want these to be removed later--they are just for debugging during development.